src/result_set.rs

```diff
@@ -492,6 +577,67 @@ impl FromRedisValueWithGraph for Relation {
     }
 }
 
+/// A path returned by RedisGraph.
```

Comment:
I think we should document how exactly the `nodes` and `edges` lists are laid out in the graph.

Reply:
This should be the same as all Cypher implementations (i.e. outside of RedisGraph): nodes and edges are returned in the order they are queried. I extended the `test_path` test to demonstrate multiple nodes/edges. For example,

```
CREATE (:L1 {prop: 1})-[:R1 {prop: 2}]->(:L2 {prop: 3})-[:R2 {prop: 4}]->(:L3 {prop: 5})
CREATE (:L1 {prop: 11})-[:R1 {prop: 22}]->(:L2 {prop: 33})-[:R2 {prop: 44}]->(:L3 {prop: 55})

MATCH p = (:L1)-[:R1]->(:L2)-[:R2]->(:L3) RETURN p
```

Returns a `Vec<Path>`, each with 3 nodes and 2 relations as specified in the query order:
```
[
	Path {
	  nodes: [ (:L1 {prop: 1}), (:L2 {prop: 3}), (:L3 {prop: 5}) ],
	  edges: [ [:R1 {prop: 2}], [:R2 {prop: 4}] ]
	},
	Path {
	  nodes: [ (:L1 {prop: 11}), (:L2 {prop: 33}), (:L3 {prop: 55}) ],
	  edges: [ [:R1 {prop: 22}], [:R2 {prop: 44}] ]
	}
]
```

Reply:
Do you think there is anything that should be explicitely documented in Rust? Other libraries don't seem to go in-depth for this, but I'd be happy to word what I just wrote above in the doc.

Reply:
Actually I just uncovered a bug in my implementation when two nodes have the same edge type twice, for example:

```
CREATE (:L1 {prop: 1})-[:R1 {prop: 2}]->(:L2 {prop: 3})
MATCH (a:L1), (b:L2) CREATE (a)-[:R1 {prop: 22}]->(b)
MATCH p = (:L1)-[:R1]->(:L2)-[:R2]->(:L3) RETURN p
```

I'll have to do a bit more digging...

Reply:
I see. From a user perspective, wouldn't it be a little strange to work with a "path" which nodes and edges have been taken apart? Why not do something like

```rust
enum Path {
    Cons(Node, Edge, Path),
    End(Node, Edge, Node),
}
```
(actually not sure if this is valid Rust, I haven't used that language in some time)

This would allow you to "walk along the path" just by doing recursion. What do you think?

Reply:
This would seem more like a job for an iterator implementation, I'll experiment with something :smiley:

Reply:
How's this?

```rust
/// A path returned by RedisGraph.
#[derive(Debug, Clone, PartialEq)]
pub struct Path {
    /// Nodes in the path.
    pub nodes: Vec<Node>,
    /// Edges in the path.
    pub edges: Vec<Edge>,
}

impl Path {
    pub fn traverse(&self) -> PathTraversal {
        PathTraversal { path: self, idx: 0 }
    }
}

pub struct PathTraversal<'a> {
    path: &'a Path,
    idx: usize,
}

impl<'a> Iterator for PathTraversal<'a> {
    type Item = (&'a Node, Option<&'a Edge>, Option<&'a Node>);

    fn next(&mut self) -> Option<Self::Item> {
        if let Some(node) = self.path.nodes.get(self.idx) {
            let next_edge = self.path.edges.get(self.idx);
            let next_node = self.path.nodes.get(self.idx + 1);
            self.idx += 1;
            Some((node, next_edge, next_node))
        } else {
            None
        }
    }
}
```

Which means you can traverse the path using:

```rust
path.traverse().for_each(|(node, edge, next_node)| {
    // ...
})
```

Or even

```rust
for (node, edge, next_node) in path.traverse() {
    // ...
}
```

For example, implementing `Display` for `Path`:
```rust
impl std::fmt::Display for Path {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        for (node, edge, _) in self.traverse() {
            write!(f, "({:?})", node)?;
            if let Some(edge) = edge {
                write!(f, "-[{:?}]->", edge)?;
            }
        }
        Ok(())
    }
}
```

If you like the general idea I'll push it for review.

Reply:
Two problems I see with this implemetation:

- Having two Vecs tells you nothing about the actual data structure. Most inhabitants of `Path` are actually not valid paths.
- The `Option<Edge>` seems unnecessarily bureaucratic to me. This is just a problem that follows from the invalid inhabitants of `Path`.

Can you use the `Path` type from my previous comment (I think you need to `Box` the tail on `Cons`, though)? I would gladly merge that.

Reply:
The `Path` struct represents how the struct is returned by RedisGraph. This is standard for all Cypher implementations: `Path` is composed of `n` edges and `n+1` nodes in sorted order. This is a common API across not just RedisGraph clients, but Cypher clients in general. [[1]](https://github.com/neo4j/neo4j-java-driver/blob/e993059737eab4ce9ca9c827274c059ba5cb8908/driver/src/main/java/org/neo4j/driver/types/Path.java) [[2]](https://github.com/RedisGraph/redisgraph-py/blob/94e263196055a7aa2dd85d41e7edc9f4561f2813/redisgraph/path.py) [[3]](https://github.com/neo4j/neo4j-python-driver/blob/8f708fe9374f7e6bdfdf84ff611f266a8a5e9344/neo4j/graph/__init__.py#L272)

Now, the *traversal* of the path is linear and could be implemented in different ways. The design I made has the assumption that a `Path` could have just 1 node and no edges. However, if the conclusion can be made that a `Path` always has at least 2 nodes and 1 edge, I could rewrite my `traverse()` iterator like this, which more closely resembles how you would implement traversal with your enum:

```rust
/// A path returned by RedisGraph.
#[derive(Debug, Clone, PartialEq)]
pub struct Path {
    /// Nodes in the path.
    pub nodes: Vec<Node>,
    /// Edges in the path.
    pub edges: Vec<Edge>,
}

impl Path {
    /// The length of the path. This is effectively the amount of [`Edge`]s in
    /// the path.
    pub fn len(&self) -> usize {
        self.edges.len()
    }

    /// Traverses the path by returning the first segment of the path, allowing
    /// recursion to traverse the entire path.
    pub fn traverse(&self) -> PathSegment {
        let len = self.len();
        let mut segment = PathSegment::End(&self.nodes[len - 1], &self.edges[len - 1], &self.nodes[len]);
        for i in (len - 2)..=0 {
            segment = PathSegment::Cons(&self.nodes[i], &self.edges[i], Box::new(segment));
        }
        segment
    }
}

impl std::fmt::Display for Path {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.traverse())
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum PathSegment<'a> {
    Cons(&'a Node, &'a Edge, Box<PathSegment<'a>>),
    End(&'a Node, &'a Edge, &'a Node),
}

impl<'a> std::fmt::Display for PathSegment<'a> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            PathSegment::Cons(start, relation, next) => {
                write!(f, "()-[]->{}", next)
            },
            PathSegment::End(start, relation, end) => {
                write!(f, "()-[]->()")
            }
        }
    }
}
```

Now the `Path` struct could be entirely replaced with `PathSegment` by converting the 2 Vecs at deserialization (in `from_redis_value_with_graph`). However, it is my opinion that the 2 Vecs-structure should remain accessible to users to cover use-cases such as:

* Getting the `i`-th node/edge without iterating through potentially the entire path
* Getting the length of the path

Reply:
I think it's a good idea to allow using both representations. How about we have a `RawPath` with two Vecs and a list-like `Path`? Then we could implement `FromCell` for both and let `Path` own its nodes and edges. That way, the user can decide which representation to use on a per query basis. Conversion and length functions could also be useful.